doc/cprover-manual/goto-instrument.md

```diff
@@ -0,0 +1,127 @@
+[CPROVER Manual TOC](../../)
+
+## Goto Instrument
+
+This is a tool that implements a variety of instrumentation passes modifying a
+pre-existing GOTO binary, or rather producing the modified binary as output.
+
+```sh
+$ goto-instrument in.gb out.gb <instrumentation-options>
+```
+
+### Diagnosis Options
+
+### Safety Checks
+
+### Semantic Transformations
+
+#### Generate Function Body
+
+This transformation inserts implementation of functions without definition, i.e.
+a body. The user is allowed to choose the behavior to be implemented by those
+functions from a predefined list. The command-line option to select the behavior
+is `--generate-function-body-options <option>`.
+
+- `assert-false`: The body consists of a single command: `assert(0);`
+- `assume-false`: The body consists of a single command: `assume(0);`
+- `nondet-return`: The body only returns a non-deterministic value of it's
+  return type.
+- `assert-false-assume-false`: Two commands as above.
+- `havoc[,params:<p-regex>][,globals:<g-regex>]`: Assign non-deterministic
+  values to the targets of pointer-to-non-constant parameters matching the
+  `p-regex`, and non-constant globals matching the `g-regex` and then (in case
+  of non-void function) returning as with `nondet-return`.
+
+Let's demonstrate using the havoc on a simple example:
+
+```C
+// main.c
+int global;
+const int c_global;
+
+int function(int *param, const int *c_param);
+```
+
+Often we want to avoid overwriting internal symbols, i.e. those with `__`
+prefix, which is done using the pattern `(?!__)`.
+
+```sh
+$ goto-cc main.c -o main.gb
+$ goto-instrument main.gb main-out.gb \
+  --generate-function-body-options 'havoc,params:(?!__).*,globals:(?!__).*' \
+  --generate-funtion-body function
+```
+
+Leads to a binary equivalent to the following C code:
+
+```C
+// main-mod.c
+int function(int *param, const int *c_param) {
+  *param = nondet_int();
+  global = nondet_int();
+  return nondet_int();
+}
+```
+
+Which parameters should be modified can be specified either by a regular
+expression (as above) or by a semicolon-separated list of their numbers. For
+example `havoc,params:0;3;4` will assign non-deterministic values to the 1st,
+4th, and 5th parameter.
+
+Finally, the user can also request different implementation for a number of
+call-sites of a single function. The option `--generate-havocing-body` inserts
```

Comment:
Ah, that makes more sense then! But isn’t the name `generate-havocing-body` a bit misleading then? I don’t know.